tools/shared.py

```diff
@@ -771,6 +771,20 @@ def exe_suffix(cmd):
   return cmd + '.exe' if WINDOWS else cmd
 
 
+def replace_suffix(filename, new_suffix):
+  if not new_suffix.startswith('.'):
+    new_suffix = '.' + new_suffix
+  return os.path.splitext(filename)[0] + new_suffix
+
+
+# In MINIMAL_RUNTIME mode, keep suffixes of generated files simple ('.mem' instead of '.js.mem'; .'symbols' instead of '.js.symbols' etc)
+# Retain the original naming scheme in traditional runtime.
+def replace_or_append_suffix(filename, new_suffix):
+  if not new_suffix.startswith('.'):
+    new_suffix = '.' + new_suffix
+  return replace_suffix(filename, new_suffix) if Settings.MINIMAL_RUNTIME else filename + new_suffix
```

Comment:
I love this PR but I dislike this line. As I said in the past, let's do this in all modes: that reduces complexity overall and makes it easier to use `MINIMAL_RUNTIME`.

The downside to doing it in all modes is it's a breaking change, but it's an obvious (no weird bugs) and beneficial one, so it's worth the minor one-time inconvenience to users I think.

Reply:
I don't agree here, making this change unconditionally makes it much harder for build systems to support multiple Emscripten versions, since they need to silently be aware that in Emscripten tag 1.x.y they decided to change what files get generated. Everyone's build systems will be impacted, but also custom shells are impacted, and there is no good way to keep supporting both, unless developers implement postprocessing options that check `emcc --version`, parse the result string, and then postprocess their shell files/html loaders depending on which version is seen, and make build systems conditional to `emcc --version`. Not many developers have this kind of machinery available and it may mean a large amount of development, especially due to the change being quite cosmetic, rather than anything important.

Pairing this change with a new option rather than breaking in version X makes it much simpler for build systems to adapt, because they can make a choice to toggle MINIMAL_RUNTIME to enabled, and then associated to that, adjust the build outputs, without having to build a feedback loop of parsing `emcc --version` in the build system. Build systems/projects that do not care about MINIMAL_RUNTIME are not impacted and they can continue to do what they were doing before. Breaking here is unnecessary, and again, the complexity here is contained in one `if()` in one function, as opposed to a compiler wide semantic across a lot of code.

Reply:
Because migrating to MINIMAL_RUNTIME needs developing a new shell file anyways, breaking the file naming scheme migrating to that is more natural, as one needs to write the file again anyways. So pairing this change to MINIMAL_RUNTIME is not as arbitrary as it superficially might look, since developer needs to revamp the shell file for MINIMAL_RUNTIME anyways.

Reply:
Fair points, but on the other hand, that's not consistent with what we've done in the past, like when we renamed the js worker file extension.

How many of our users support multiple emscripten versions at once? Perhaps worth asking on the mailing list.

Reply:
Or build system at Unity does, and that build system is used to build different projects, which update Emscripten at different cadence. There is a separate build system team, and project teams, so making this orthogonal lets one team migrate to new Emscripten with MINIMAL_RUNTIME, while letting the other team stick to the other version, until we are ready to update that project as well.

In general it might not be a question of "how manu users support multiple emscripten versions at once", but rather a question of "how many build systems need to support multiple emscripten versions at once", to which the question is probably "all of them". Build systems often need to codify in them the set of build inputs and build outputs to build a build dependency graph. Requiring that the build system does `emcc --version` test invocations before starting any work to build a knowledge base about which output files will be present is real complexity. Having that choice come from an input passed to Emscripten link stage makes this trivial one-liner to support in a build system - it can do an easy check
```
memOutputFile = 'MINIMAL_RUNTIME' in linkerFlags ? unsuffixed(outputName) + '.mem' : outputName + '.mem';
```

If this is too awkward to accommodate in Emscripten, then I don't know anything else than to maintain our own fork that lets us keep a migration story, or drop this PR and revert the other .mem file name change that already landed, as too complex naming change to achieve.

Changing the JS worker file and the Fetch file were not issues that we'd yell about, because multithreading is still in infancy, and we did not have to support two Emscripten versions both with multithreading support at the same time. If we had, it would have been a problem (for us), and we would have yelled. I can imagine an unconditional change of a.js.mem to a.mem and a.js.symbols to a.symbols etc. for all uses would make many people yell, which is why I guarded this against MINIMAL_RUNTIME to make sure nobody will need to yell.

Reply:
Is it hard to make the build system do "copy over `a.js.mem` if it's there, and copy `a.mem` if it's there (or both, but that would never happen)"? I feel like I don't understand the real issue here yet.

Yes, this can cause some inconvenience, I appreciate that, but it seems manageable?

Reply:
At least in Unity build system unfortunately not, because those files are outputs, and the build graph checks the existence and timestamps of the outputs to decide if it needs to build or not, so it would need to know which file it is supposed to be stat()ting for the timestamps, already before it has run any build commands.

Reply:
I'd like to make sure I understand what are you suggesting here. If it is that the minimal and non-minimal cases stay different forever, or is there a migration story here, and at some point in the future you would not object to this breaking change? For example, do you intend to switch entirely to the minimal runtime at some point, and so not care about the other mode?

Reply:
Yeah, I am proposing that they'd stay different forever. To my mind the grand claim is still that everyone will eventually move to MINIMAL_RUNTIME as the only runtime, when it encompasses all the relevant uses that people have (and the rest could be dropped as legacy, such as library_browser.js), at which point the old naming scheme along with the old runtime would become obsolete, so that'd be a natural migration. (though I don't know how long that will take, or if there will end up being pain points that prevent such migration)

> at some point in the future you would not object to this breaking change?

Subjectively speaking, once both our projects have been converted over to MINIMAL_RUNTIME, we do not care about breakages in the traditional runtime. (To fully convert over to MINIMAL_RUNTIME, we need to do exports support in minimal runtime, and multithreading support, and fetch+indexeddb+asmfs filesystem support that come to mind at least)

Reply:
Interesting,  I think within our build system we would have the opposite problem to you.  We have at least one post-link step that looks for the build output files by name.  This step knows which emscripten version is in use (it versions with emscripten actually), but it does not have access to the link flags the user chose.  So it would be strictly worse for our build system if the linker flags effected the names of these output files.  I suppose we could pass the users link flags to the post build step but that seem rather cumbersome.

In general I don't see how a build system that supports multiple emcc versions is able to function *without* knowing the version it is currently dealing with.  How can it make decisions like "can i use feature X", and "should I pass flag Y" without knowing the version.  For example, I'm trying to make a change right now so that "-flto" can be used to enable LTO, but that won't work in older versions of emscripten so any build system that supports multiple versions would presumably need to say something like"

```
if project_selected_lto_feature:
  if EMCC_VERSION > X.Y:
    use_flto_flag()
  else:
    do_something_else()
```

Even for native toolchains like clang you need to know the exact version you are using, for example, to know if you can enable a new class of warning with "-Wfoo". 

It does seem like there are two different philosophy at play here based on this discussion and in https://github.com/emscripten-core/emscripten/pull/8034.

In one we allow emscripten to evolve in backwards incompatible ways, especially if it means we can reduce our technical debt and simplify our code.   The other is that we should continue to maintain command line compatibility across versions even if we have to jump through hoops internally to achieve this.

Admittedly the former approach does force developers to be aware of changes we make as they perform toolchain upgrades.

Reply:
> This step knows which emscripten version is in use (it versions with emscripten actually), but it does not have access to the link flags the user chose. So it would be strictly worse for our build system if the linker flags effected the names of these output files.

This is already what happens all over throughout Emscripten, the linker flags affect the names of the output files, and not the Emscripten version:
- if you build with linker flag `-s WASM=1`, you get a file `a.wasm`,
- if you build with linker flag `-s WASM=0`, you will not get a file `a.wasm`,
- if you build with linker flag `-s WASM=0 --separate-asm`, you will get a file `a.asm.js`,
- if you build with linker flag `--preload-file x`, you will get a file `a.data`,
- if you build with linker flag `--memory-init-file 1`, you will get a file `a(.html).mem`,
- if you build with linker flag `-s USE_PTHREADS=1`, you will get a file `a.worker.js`,
- if you build with linker flag `-s USE_PTHREADS=1 -s FETCH=1`, you will get a file `a.fetch.js`,
- if you build with linker flag `--emit-symbol-map`, you will get a file `a(.html).symbols`,
- if you build with a linker flag `-s SINGLE_FILE=1`, you will not get `a.js` or `a.asm.js` or `a.data` or `a.mem`, but you will just get `a.html`.

None of these decisions are gated on which Emscripten version one is using.

If you are saying that your build system does not look at the linker settings passed to Emscripten when it is determining what output files are being generated, then your build system is extremely narrow in its supported/capable use cases, or just buggy. When you say your build system would have the "opposite" problem, I do not believe that to be correct.

The proposal to adjust dropping the `(.html)` parts from the above files when `-s MINIMAL_RUNTIME=1` is passed plays well together with existing code in a build system that already has to have the code to determine what the set of output files is, based on linker settings that are passed.

> How can it make decisions like "can i use feature X"

There are three components in play: a build system, the compiler, and the project that you are building. You are making a fallacy here that the build system would dictate which version of the compiler to use for a project, whereas this is not the case in practically all uses of a build system, but it is the project that decides which version of the compiler to pair with the project. That is, the right question is not to ask "how can a build system make decisions like "can i use feature X", but to ask "how does a project make decisions like "can i use feature X".

For example in the case of -s MINIMAL_RUNTIME=1, it is not the build system that makes a decision when to use it or not, but it is the project you are developing that decides when to use the minimal runtime. To switch over, the project needs to update the compiler at the same time to get to the compiler that supports the feature, and then the project knows it will need to pass that flag to tell the build system.

> For example, I'm trying to make a change right now so that "-flto" can be used to enable LTO, but that won't work in older versions of emscripten so any build system that supports multiple versions would presumably need to say something like"

No, the build system does not have to say anything. If proper compiler development practices with backwards compatibility have been followed, the build system does not have to complicate things for itself. Project enables LTO by signalling to the build system that LTO needs to be enabled. Either the project says to the build system `ldflags += -flto`, or the project says to the build system `project.linker.enableLto()` (depending on how much build system wants to know). In second case, build system, if it has been updated to support new compiler, passes `ldflags += -flto` since it is now aware of the setting. If this scheme is attempted on an old Emscripten version, then Emscripten will yell about unknown flag `-flto` being seen, and abort the build (and not the build system). The build system does not need to bake this validation into it, because it would just be overvalidating the compiler for the joy of overvalidating and no real purpose.

I am not claiming that there can not be scenarios where a build system did not need to start gating items on Emscripten versions, but such scenarios are rare, and generally occur as result of backwards compatibility practices having gone south in compiler development itself.

> It does seem like there are two different philosophy at play here based on this discussion and in #8034.

I do not think this is a question of philosophy, but a question of basic hygiene with compiler development. By ensuring proper compatibility story, we can orthogonally/in parallel update our build system to support the new compiler, and update the compiler we host in our internal CDN, and projects can update the build system, and the compiler independently of each other.

If this rename is done as a breaking change, existing build systems become useless and they will need a fix to keep supporting new compiler, and projects will be locked out from updating to a new compiler until they get a fixed build system. The build system side fix will need to contain awkward plumbing in order to maintain support for old compilers, in case a project might want to update just the build system (for other reasons), but not the compiler.

If this rename is done with proper hygiene, gating on a linker setting like other Emscripten features already do, then projects can choose to orthogonally update their build system, or their compiler, and then migrate to the new feature when they have been able to update both and are ready for the new feature.